internal/handlers/common/projection.go

```diff
@@ -26,78 +26,148 @@ import (
 	"github.com/FerretDB/FerretDB/internal/util/must"
 )
 
-// isProjectionInclusion: projection can be only inclusion or exclusion. Validate and return true if inclusion.
+const (
+	projectionElemMatch = "$elemMatch"
+	projectionSlice     = "$slice"
+)
+
+// validateProjectionExpression: projection can be only inclusion or exclusion. Validate and return true if inclusion.
 // Exception for the _id field.
-func isProjectionInclusion(projection *types.Document) (inclusion bool, err error) {
-	var exclusion bool
+func validateProjectionExpression(projection *types.Document) (bool, error) {
+	inclusion, _, err := validateExpression(projection, 0, false, false)
+	return inclusion, err
+}
+
+func validateExpression(projection *types.Document, depth int, inclusion, exclusion bool) (bool, bool, error) {
+	var err error
 	for _, k := range projection.Keys() {
 		if k == "_id" { // _id is a special case and can be both
 			continue
 		}
+
 		v := must.NotFail(projection.Get(k))
 		switch v := v.(type) {
 		case *types.Document:
-			for _, projectionType := range v.Keys() {
-				supportedProjectionTypes := []string{"$elemMatch", "$slice"}
-				if !slices.Contains(supportedProjectionTypes, projectionType) {
-					err = lazyerrors.Errorf("projection of %s is not supported", projectionType)
-					return
-				}
 
-				switch projectionType {
-				case "$elemMatch":
-					inclusion = true
-				case "$slice":
-					inclusion = false
-				default:
-					panic(projectionType + " not supported")
-				}
+			inclusion, exclusion, err = validateDocProjectionExpression(v, depth, inclusion, exclusion)
+
+		case *types.Array:
+			err = NewErrorMsg(ErrElemMatchObjectRequired, "elemMatch: Invalid argument, object required, but got array")
+			return false, false, err
+
+		default: // scalar
+			if k == projectionElemMatch {
+				err = NewError(ErrElemMatchObjectRequired,
+					fmt.Errorf("elemMatch: Invalid argument, object required, but got %T", v),
+				)
+				return false, false, err
 			}
 
-		case float64, int32, int64:
-			if types.Compare(v, int32(0)) == types.Equal {
-				if inclusion {
-					err = NewError(ErrProjectionExIn,
-						fmt.Errorf("Cannot do exclusion on field %s in inclusion projection", k),
-					)
-					return
-				}
-				exclusion = true
-			} else {
-				if exclusion {
-					err = NewError(ErrProjectionInEx,
-						fmt.Errorf("Cannot do inclusion on field %s in exclusion projection", k),
-					)
-					return
-				}
-				inclusion = true
+			inclusion, exclusion, err = validateScalarProjectionExpression(v, k, inclusion, exclusion)
+		}
+	}
+	return inclusion, exclusion, err
+}
+
+func validateScalarProjectionExpression(v any, field string, inclusion, exclusion bool) (bool, bool, error) {
+	var err error
```

Comment:
You don't need this variable as you set the error and return its value in all cases.
And in the last return, you could just use `nil`.

Reply:
I don't want to make `:=` each time

Reply:
whenever it was declared it is nil, if it's initialized it's returned